File: client/src/repositories/taskRepository.ts
import { db, LocalTask, LocalClient } from '@/db/schema';
import { enqueueOutbox, cancelPendingCreate } from '@/db/outbox';
import { Task, TaskWithClient, TaskClientRef, Priority, EntityStatus } from '@/types';
import { TaskListParams, TaskCreatePayload } from '@/services/taskService';
import { defaultTaskSort } from '@/utils/taskSort';
import { isExpiredCompletedTask } from '@/utils/formatters';

function toTask(row: LocalTask): Task {
  return {
    ...row,
    createdAt: row.createdAt.toISOString(),
    updatedAt: row.updatedAt.toISOString(),
    completedAt: row.completedAt?.toISOString(),
    dueDate: row.dueDate?.toISOString(),
  };
}

function toLocalTask(task: Task): LocalTask {
  return {
    ...task,
    createdAt: new Date(task.createdAt),
    updatedAt: new Date(task.updatedAt),
    completedAt: task.completedAt ? new Date(task.completedAt) : undefined,
    dueDate: task.dueDate ? new Date(task.dueDate) : undefined,
  };
}

function toClientRef(row: LocalClient): TaskClientRef {
  return { id: row.id, name: row.name, avatarUrl: row.avatarUrl, initials: row.initials };
}

async function attachClients(tasks: Task[]): Promise<TaskWithClient[]> {
  const ids = [...new Set(tasks.map((t) => t.clientId).filter((id): id is string => Boolean(id)))];
  if (ids.length === 0) return tasks;

  const clients = await db.clients.bulkGet(ids);
  const byId = new Map(clients.filter((c): c is LocalClient => Boolean(c)).map((c) => [c.id, toClientRef(c)]));

  return tasks.map((task) => ({ ...task, client: task.clientId ? byId.get(task.clientId) : undefined }));
}

export interface TaskFormInput {
  title: string;
  description?: string;
  clientId?: string;
  dueDate?: string;
  priority: Priority;
  status?: EntityStatus;
  reminderEnabled?: boolean;
}

async function list(params: TaskListParams = {}): Promise<TaskWithClient[]> {
  let rows = await db.tasks.toArray();

  // Completed tasks past their 24h stay are already on their way out (see
  // purgeExpiredCompleted, which runs on the sync cycle). Dropping them here
  // too means the list never shows one in the window between expiring and
  // the next purge pass.
  rows = rows.filter((r) => !isExpiredCompletedTask(r));

  if (params.status) rows = rows.filter((r) => r.status === params.status);
  if (params.priority) rows = rows.filter((r) => r.priority === params.priority);
  if (params.clientId) rows = rows.filter((r) => r.clientId === params.clientId);
  if (params.search) {
    const term = params.search.trim().toLowerCase();
    rows = rows.filter(
      (r) => r.title.toLowerCase().includes(term) || (r.description ?? '').toLowerCase().includes(term)
    );
  }

  let tasks = rows.map(toTask);

  if (params.sort) {
    const order = params.order === 'asc' ? 1 : -1;
    tasks = [...tasks].sort((a, b) => {
      const av = a[params.sort as keyof Task];
      const bv = b[params.sort as keyof Task];
      if (av === bv) return 0;
      if (av === undefined) return 1;
      if (bv === undefined) return -1;
      return av > bv ? order : -order;
    });
  } else {
    tasks = [...tasks].sort(defaultTaskSort);
  }

  return attachClients(tasks);
}

async function get(id: string): Promise<TaskWithClient | undefined> {
  const row = await db.tasks.get(id);
  if (!row) return undefined;
  const [withClient] = await attachClients([toTask(row)]);
  return withClient;
}

async function create(input: TaskFormInput): Promise<TaskWithClient> {
  const now = new Date();
  const status = input.status ?? 'pending';
  const row: LocalTask = {
    id: crypto.randomUUID(),
    title: input.title,
    description: input.description,
    clientId: input.clientId || undefined,
    dueDate: input.dueDate ? new Date(input.dueDate) : undefined,
    priority: input.priority,
    status,
    reminderEnabled: input.reminderEnabled ?? false,
    createdAt: now,
    updatedAt: now,
    // A task filed as already done starts its 24h retention right away —
    // completedAt has to be set wherever status becomes 'completed', not
    // only in updateStatus, or the row would never age out.
    completedAt: status === 'completed' ? now : undefined,
  };

  await db.tasks.put(row);

  const payload: TaskCreatePayload = {
    localId: row.id,
    title: row.title,
    description: row.description,
    clientId: row.clientId,
    dueDate: row.dueDate?.toISOString(),
    priority: row.priority,
    status: row.status,
    reminderEnabled: row.reminderEnabled,
    createdAt: row.createdAt.toISOString(),
    updatedAt: row.updatedAt.toISOString(),
    completedAt: row.completedAt?.toISOString(),
  };
  await enqueueOutbox('task', row.id, 'create', payload);

  const [withClient] = await attachClients([toTask(row)]);
  return withClient;
}

async function update(id: string, input: Partial<TaskFormInput>): Promise<TaskWithClient> {
  const existing = await db.tasks.get(id);
  if (!existing) throw new Error('Tarefa não encontrada localmente.');

  const now = new Date();
  const status = input.status ?? existing.status;
  const row: LocalTask = {
    ...existing,
    ...input,
    dueDate: input.dueDate !== undefined ? (input.dueDate ? new Date(input.dueDate) : undefined) : existing.dueDate,
    clientId: 'clientId' in input ? input.clientId || undefined : existing.clientId,
    updatedAt: now,
    // The edit form can flip status just as the checkbox does, so keep the
    // completion stamp in step here too — it is what the 24h retention
    // counts from. Re-completing an already-done task keeps the original
    // stamp rather than restarting the clock.
    completedAt: status === 'completed' ? existing.completedAt ?? now : undefined,
  };

  await db.tasks.put(row);
  await enqueueOutbox('task', id, 'update', {
    title: row.title,
    description: row.description,
    clientId: row.clientId ?? '',
    dueDate: row.dueDate?.toISOString(),
    priority: row.priority,
    status: row.status,
    reminderEnabled: row.reminderEnabled ?? false,
    updatedAt: row.updatedAt.toISOString(),
    completedAt: row.completedAt?.toISOString(),
  });

  const [withClient] = await attachClients([toTask(row)]);
  return withClient;
}

/** Quick opt-in/out toggle for the "sininho" reminder — reuses update() so
 * it rides the same outbox/sync path as every other field edit. */
async function setReminder(id: string, enabled: boolean): Promise<TaskWithClient> {
  return update(id, { reminderEnabled: enabled });
}

async function updateStatus(id: string, status: EntityStatus): Promise<TaskWithClient> {
  const existing = await db.tasks.get(id);
  if (!existing) throw new Error('Tarefa não encontrada localmente.');

  const now = new Date();
  const completedAt = status === 'completed' ? now : undefined;
  const row: LocalTask = { ...existing, status, updatedAt: now, completedAt };

  await db.tasks.put(row);
  await enqueueOutbox('task', id, 'status', {
    status,
    updatedAt: now.toISOString(),
    completedAt: completedAt?.toISOString(),
  });

  const [withClient] = await attachClients([toTask(row)]);
  return withClient;
}

/**
 * Drops completed tasks whose 24h stay has run out — from IndexedDB now and,
 * through the outbox, from the server on the next push. Called at the top of
 * every sync cycle, so it also runs offline: the local row goes immediately
 * and the delete rides out on reconnect.
 *
 * Returns how many rows it removed.
 */
async function purgeExpiredCompleted(): Promise<number> {
  const [completed, queued] = await Promise.all([
    db.tasks.where('status').equals('completed').toArray(),
    db.outbox.where('entity').equals('task').toArray(),
  ]);

  // A pull can hand a row back before its queued delete has made it to the
  // server; re-queueing the same delete every cycle would just grow the
  // outbox, so leave those to the entry that's already waiting.
  const awaitingDelete = new Set(
    queued.filter((entry) => entry.type === 'delete').map((entry) => entry.entityId)
  );
  const expired = completed.filter((row) => !awaitingDelete.has(row.id) && isExpiredCompletedTask(row));

  for (const row of expired) {
    await remove(row.id);
  }

  return expired.length;
}

async function remove(id: string): Promise<void> {
  await db.tasks.delete(id);
  const cancelled = await cancelPendingCreate('task', id);
  if (!cancelled) {
    await enqueueOutbox('task', id, 'delete');
  }
}

async function cascadeClientRemoval(clientId: string, action: 'unlink' | 'delete'): Promise<void> {
  const affected = await db.tasks.where('clientId').equals(clientId).toArray();

  for (const task of affected) {
    if (action === 'delete') {
      await remove(task.id);
    } else {
      await update(task.id, { clientId: undefined });
    }
  }
}

async function upsertFromServer(task: Task): Promise<void> {
  const existing = await db.tasks.get(task.id);
  if (existing && existing.updatedAt.toISOString() > task.updatedAt) return;
  await db.tasks.put(toLocalTask(task));
}

async function replaceLocal(task: Task): Promise<void> {
  await db.tasks.put(toLocalTask(task));
}

async function getAllLocalIds(): Promise<Set<string>> {
  const ids = await db.tasks.toCollection().primaryKeys();
  return new Set(ids);
}

async function removeLocalOnly(id: string): Promise<void> {
  await db.tasks.delete(id);
}

export const taskRepository = {
  list,
  get,
  create,
  update,
  updateStatus,
  setReminder,
  purgeExpiredCompleted,
  remove,
  cascadeClientRemoval,
  upsertFromServer,
  replaceLocal,
  getAllLocalIds,
  removeLocalOnly,
};

File: flutter_app/lib/data/repositories/task_repository.dart
import 'package:uuid/uuid.dart';

import '../../models/enums.dart';
import '../../models/models.dart';
import '../../utils/formatters.dart';
import '../local/local_db.dart';
import '../local/outbox.dart';
import '../remote/services.dart';

const _uuid = Uuid();

/// Mirrors server/src/utils/taskSort.ts — the list must read the same online
/// or off: overdue first, then priority, then due date, then newest.
int defaultTaskSort(Task a, Task b) {
  final overdueA = isOverdue(a.dueDate, a.status) ? 0 : 1;
  final overdueB = isOverdue(b.dueDate, b.status) ? 0 : 1;
  if (overdueA != overdueB) return overdueA - overdueB;

  final priorityDiff = a.priority.rank - b.priority.rank;
  if (priorityDiff != 0) return priorityDiff;

  final dueA = a.dueDate?.millisecondsSinceEpoch ?? double.maxFinite.toInt();
  final dueB = b.dueDate?.millisecondsSinceEpoch ?? double.maxFinite.toInt();
  if (dueA != dueB) return dueA.compareTo(dueB);

  return b.createdAt.compareTo(a.createdAt);
}

class TaskFormInput {
  const TaskFormInput({
    required this.title,
    required this.priority,
    this.description,
    this.clientId,
    this.dueDate,
    this.status,
    this.reminderEnabled,
    this.clearDueDate = false,
    this.clearClientId = false,
  });

  final String title;
  final Priority priority;
  final String? description;
  final String? clientId;
  final DateTime? dueDate;
  final EntityStatus? status;
  final bool? reminderEnabled;
  final bool clearDueDate;
  final bool clearClientId;
}

class TaskQuery {
  const TaskQuery({this.search, this.status, this.priority, this.clientId, this.sort, this.ascending = true});
  final String? search;
  final EntityStatus? status;
  final Priority? priority;
  final String? clientId;
  final String? sort;
  final bool ascending;
}

class TaskRepository {
  TaskRepository(this._db) : _outbox = Outbox(_db);

  final LocalDb _db;
  final Outbox _outbox;

  List<Task> _all() => allJson(_db.tasks).map(Task.fromJson).toList();

  /// Joins the small bits of the Client the UI needs. Resolved locally, never
  /// fetched — it has to work offline.
  List<Task> _attachClients(List<Task> tasks) {
    final byId = <String, TaskClientRef>{};
    for (final raw in _db.clients.values) {
      final client = Client.fromJson(asJson(raw));
      byId[client.id] = TaskClientRef.fromClient(client);
    }
    return tasks
        .map((task) => task.clientId == null ? task : task.copyWith(client: byId[task.clientId]))
        .toList();
  }

  List<Task> list([TaskQuery query = const TaskQuery()]) {
    var rows = _all();

    // Completed tasks past their 24h stay are already on their way out (see
    // purgeExpiredCompleted, which runs on the sync cycle). Dropping them
    // here too means the list never shows one in the window between expiring
    // and the next purge pass.
    rows = rows.where((t) => !_isExpired(t)).toList();

    if (query.status != null) rows = rows.where((t) => t.status == query.status).toList();
    if (query.priority != null) rows = rows.where((t) => t.priority == query.priority).toList();
    if (query.clientId != null) rows = rows.where((t) => t.clientId == query.clientId).toList();
    if (query.search != null && query.search!.trim().isNotEmpty) {
      final term = query.search!.trim().toLowerCase();
      rows = rows
          .where((t) =>
              t.title.toLowerCase().contains(term) ||
              (t.description ?? '').toLowerCase().contains(term))
          .toList();
    }

    if (query.sort == null) {
      rows.sort(defaultTaskSort);
    } else {
      final direction = query.ascending ? 1 : -1;
      rows.sort((a, b) => switch (query.sort) {
            'dueDate' => direction * _compareNullableDates(a.dueDate, b.dueDate),
            'priority' => direction * a.priority.rank.compareTo(b.priority.rank),
            'status' => direction * a.status.index.compareTo(b.status.index),
            _ => direction * a.createdAt.compareTo(b.createdAt),
          });
    }

    return _attachClients(rows);
  }

  Task? get(String id) {
    final raw = _db.tasks.get(id);
    if (raw == null) return null;
    return _attachClients([Task.fromJson(asJson(raw))]).first;
  }

  Future<Task> create(TaskFormInput input) async {
    final now = DateTime.now();
    final status = input.status ?? EntityStatus.pending;
    final task = Task(
      id: _uuid.v4(),
      title: input.title,
      description: input.description,
      clientId: input.clientId,
      dueDate: input.dueDate,
      priority: input.priority,
      status: status,
      reminderEnabled: input.reminderEnabled ?? false,
      createdAt: now,
      updatedAt: now,
      // A task filed as already done starts its 24h retention right away —
      // completedAt has to be set wherever status becomes completed, not only
      // in updateStatus, or the row would never age out.
      completedAt: status == EntityStatus.completed ? now : null,
    );

    await _db.tasks.put(task.id, task.toJson());
    await _outbox.enqueue(OutboxEntity.task, task.id, OutboxType.create, taskCreatePayload(task));
    return _attachClients([task]).first;
  }

  Future<Task> update(String id, TaskFormInput input) async {
    final existing = get(id);
    if (existing == null) throw StateError('Tarefa não encontrada localmente.');

    final now = DateTime.now();
    final status = input.status ?? existing.status;
    // The edit form can flip status just as the checkbox does, so keep the
    // completion stamp in step here too — it is what the 24h retention counts
    // from. Re-completing an already-done task keeps the original stamp
    // rather than restarting the clock.
    final completedAt =
        status == EntityStatus.completed ? existing.completedAt ?? now : null;

    final updated = existing.copyWith(
      title: input.title,
      description: input.description,
      clearDescription: input.description == null,
      clientId: input.clientId,
      clearClientId: input.clearClientId,
      dueDate: input.dueDate,
      clearDueDate: input.clearDueDate,
      priority: input.priority,
      status: input.status,
      reminderEnabled: input.reminderEnabled,
      updatedAt: now,
      completedAt: completedAt,
      clearCompletedAt: completedAt == null,
    );

    await _db.tasks.put(id, updated.toJson());
    await _outbox.enqueue(OutboxEntity.task, id, OutboxType.update, taskUpdatePayload(updated));
    return _attachClients([updated]).first;
  }

  /// Quick opt-in/out toggle for the reminder bell — reuses the same outbox
  /// path as any other field edit, so there is no separate branch to keep in
  /// step.
  Future<Task> setReminder(String id, bool enabled) async {
    final existing = get(id);
    if (existing == null) throw StateError('Tarefa não encontrada localmente.');

    final updated = existing.copyWith(reminderEnabled: enabled, updatedAt: DateTime.now());
    await _db.tasks.put(id, updated.toJson());
    await _outbox.enqueue(OutboxEntity.task, id, OutboxType.update, taskUpdatePayload(updated));
    return _attachClients([updated]).first;
  }

  Future<Task> updateStatus(String id, EntityStatus status) async {
    final existing = get(id);
    if (existing == null) throw StateError('Tarefa não encontrada localmente.');

    final now = DateTime.now();
    final completedAt = status == EntityStatus.completed ? now : null;
    final updated = existing.copyWith(
      status: status,
      updatedAt: now,
      completedAt: completedAt,
      clearCompletedAt: completedAt == null,
    );

    await _db.tasks.put(id, updated.toJson());
    await _outbox.enqueue(
      OutboxEntity.task,
      id,
      OutboxType.status,
      statusPayload(status, now, completedAt),
    );
    return _attachClients([updated]).first;
  }

  Future<void> remove(String id) async {
    await _db.tasks.delete(id);
    final cancelled = await _outbox.cancelPendingCreate(OutboxEntity.task, id);
    if (!cancelled) {
      await _outbox.enqueue(OutboxEntity.task, id, OutboxType.delete);
    }
  }

  bool _isExpired(Task task) => isExpiredCompletedTask(
        completed: task.status == EntityStatus.completed,
        completedAt: task.completedAt,
        updatedAt: task.updatedAt,
      );

  /// Drops completed tasks whose 24h stay has run out — from the local store
  /// now and, through the outbox, from the server on the next push. Called at
  /// the top of every sync cycle, so it also runs offline: the local row goes
  /// immediately and the delete rides out on reconnect.
  ///
  /// Returns how many rows it removed.
  Future<int> purgeExpiredCompleted() async {
    // A pull can hand a row back before its queued delete has made it to the
    // server; re-queueing the same delete every cycle would just grow the
    // outbox, so leave those to the entry that is already waiting.
    final awaitingDelete = _outbox
        .pending()
        .where((e) => e.entity == OutboxEntity.task && e.type == OutboxType.delete)
        .map((e) => e.entityId)
        .toSet();

    final expired =
        _all().where((t) => !awaitingDelete.contains(t.id) && _isExpired(t)).toList();

    for (final task in expired) {
      await remove(task.id);
    }

    return expired.length;
  }

  Future<void> cascadeClientRemoval(String clientId, String action) async {
    final affected = _all().where((t) => t.clientId == clientId).toList();

    for (final task in affected) {
      if (action == 'delete') {
        await remove(task.id);
      } else {
        final updated = task.copyWith(clearClientId: true, updatedAt: DateTime.now());
        await _db.tasks.put(task.id, updated.toJson());
        await _outbox.enqueue(OutboxEntity.task, task.id, OutboxType.update, taskUpdatePayload(updated));
      }
    }
  }

  Future<void> upsertFromServer(Task task) async {
    final raw = _db.tasks.get(task.id);
    if (raw != null && Task.fromJson(asJson(raw)).updatedAt.isAfter(task.updatedAt)) return;
    await _db.tasks.put(task.id, task.toJson());
  }

  Future<void> replaceLocal(Task task) => _db.tasks.put(task.id, task.toJson());

  Set<String> allLocalIds() => _db.tasks.keys.map((k) => k.toString()).toSet();

  Future<void> removeLocalOnly(String id) => _db.tasks.delete(id);
}

int _compareNullableDates(DateTime? a, DateTime? b) {
  if (a == null && b == null) return 0;
  if (a == null) return 1;
  if (b == null) return -1;
  return a.compareTo(b);
}
